include/cuco/detail/static_set/static_set_ref.inl

```diff
@@ -0,0 +1,300 @@
+   *
+   * @param slot Pointer to the slot in memory
+   * @param value Element to insert
+   *
+   * @return Result of this operation, i.e., success/continue/duplicate
+   */
+  __device__ inline insert_result attempt_insert(value_type* slot, value_type const& value)
+  {
+    // code path for static_set -> cas
+    if constexpr (std::is_same_v<key_type, value_type>) { return cas(slot, value); }
+
+    // TODO code path for static_map and static_multimap
+  }
+
+  /**
+   * @brief Try insert using simple atomic compare-and-swap.
+   *
+   * @param slot Pointer to the slot in memory
+   * @param value Element to insert
+   *
+   * @return Result of this operation, i.e., success/continue/duplicate
+   */
+  __device__ inline insert_result cas(value_type* slot, value_type const& value)
+  {
+    auto& ref_ = static_cast<ref_type&>(*this);
+
+    auto ref      = cuda::atomic_ref<value_type, Scope>{*slot};
+    auto expected = ref_.empty_key_sentinel_.value;
+    bool result   = ref.compare_exchange_strong(expected, value, cuda::std::memory_order_relaxed);
+    if (result) {
+      return insert_result::SUCCESS;
+    } else {
+      auto old = expected;
+      return ref_.predicate_(old, value) == detail::equal_result::EQUAL ? insert_result::DUPLICATE
+                                                                        : insert_result::CONTINUE;
+    }
+  }
+
+  // TODO packed_cas, back_to_back_cas, cas_dependent_write
+};
+
+template <typename Key,
+          cuda::thread_scope Scope,
+          typename KeyEqual,
+          typename ProbingScheme,
+          typename StorageRef,
+          typename... Operators>
+class operator_impl<op::contains_tag,
+                    static_set_ref<Key, Scope, KeyEqual, ProbingScheme, StorageRef, Operators...>> {
+  using base_type  = static_set_ref<Key, Scope, KeyEqual, ProbingScheme, StorageRef>;
+  using ref_type   = static_set_ref<Key, Scope, KeyEqual, ProbingScheme, StorageRef, Operators...>;
+  using key_type   = typename base_type::key_type;
+  using value_type = typename base_type::value_type;
+
+  static constexpr auto cg_size     = base_type::cg_size;
+  static constexpr auto window_size = base_type::window_size;
+
+ public:
+  /**
+   * @brief Indicates whether the probe key `key` was inserted into the container.
+   *
+   * If the probe key `key` was inserted into the container, returns
+   * true. Otherwise, returns false.
+   *
+   * @tparam ProbeKey Probe key type
+   *
+   * @param key The key to search for
+   * @return A boolean indicating whether the probe key is present
+   */
+  template <typename ProbeKey>
+  [[nodiscard]] __device__ inline bool contains(ProbeKey const& key) const noexcept
+  {
+    // CRTP: cast `this` to the actual reference type
+    auto const& ref_ = static_cast<ref_type const&>(*this);
+
+    auto probing_iter = ref_.probing_scheme_(key, ref_.storage_ref_.num_windows());
+
+    while (true) {
+      // TODO do we need to use atomic_ref::load here?
```

Comment:
@jrhemstad would this be needed in terms of correctness, in case a mutation operator like `cuco::insert` is enabled?

Reply:
Ideally we'd customize the implementation of `contains` such that it uses an atomic load when an `insert` operation is also present and otherwise do non-atomic loads.

Reply:
That's what I had in mind 👍

Reply:
For concurrent `insert` and `find`, I can see atomic loads are necessary but do we really need atomic loads for `insert` with `contains` or `count`?

Reply:
concurrent `insert` and `contains` is no different than concurrent `insert` and `find`.

Reply:
The bad thing about atomic loads is that we probably won't be able to use window probing as each window entry would translate to a single atomic load instruction.

We could always use relaxed loads + windows during probing and when we find a candidate slot we properly/atomically load it again to make sure we actually found a matching key. Not sure how this impacts performance.

Reply:
I can't make the atomic load thing work with the current `storage_ref` implementation. We need to refactor it first. I would suggest putting this issue in the backlog for now so we can get this PR merged since it's a blocker for other work on this branch.